src/sagemaker_xgboost_container/checkpointing.py

```diff
@@ -0,0 +1,317 @@
+    only the `max_to_keep` (5 by default) most recent checkpoints in the
+    directory. This is accomplished by a background thread that deletes all
+    checkpoints older than 5 most recent checkpoints (the number of files to
+    keep is somewhat arbitrary, choosing the optimal number of files to keep is
+    left for future work). Note that when a file is being uploaded by SM, it
+    will create a marker file (file name + .sagemaker-uploading) to indicate
+    that the file is being uploaded. SM will also create another marker file
+    (file name + .sagemaker-uploaded) when the upload is completed. Thus, the
+    background will skip deleting a file and try again later if there is a
+    marker file <filename>.sagemaker-uploading and only attempt to delete a
+    file when the marker file <filename>.sagemaker-uploaded is present.
+
+    Attributes:
+        checkpoint_dir: indicates the path to the directory where checkpoints
+            will be saved.  Defaults to /opt/ml/checkpoints on SageMaker.
+        max_to_keep: indicates the maximum number of recent checkpoint files to
+            keep.  As new files are created, older files are deleted.  Defaults
+            to 5 (that is, the 5 most recent checkpoint files are kept.)
+        start_iteration: indicates the round at which the current training
+            started. If xgb_model was loaded from a previous checkpoint, this
+            will be greater than 0 (that is, if the previous training ended
+            after round 19, start_iteration will be 20).
+        num_round: (optional) indicates the number of boosting rounds.
+
+    Example:
+        >>> save_checkpoint = SaveCheckpoint("/opt/ml/checkpoints")
+        >>> xgboost.train(prams, dtrain, callbacks=[save_checkpoint])
+    """
+    SENTINEL = None
+
+    def __init__(self, checkpoint_dir, start_iteration=0, max_to_keep=5, num_round=None):
+        """Inits SaveCheckpoint with checkpoint_dir"""
+        self.checkpoint_dir = checkpoint_dir
+        self.max_to_keep = max_to_keep
+        self.start_iteration = start_iteration
+        self.num_round = num_round
+
+        if not os.path.exists(self.checkpoint_dir):
+            os.makedirs(self.checkpoint_dir)
+        self.previous_checkpoints = [
+            os.path.join(self.checkpoint_dir, f)
+            for f in os.listdir(self.checkpoint_dir)]
+
+        self.thread = None
+        self.delete_queue = queue.Queue()
+
+        self.start()
+
+    def __call__(self, env):
+        """Makes the class callable since it is meant be used as a callback"""
+        return self.callback(env)
+
+    def format_path(self, iteration):
+        """Returns a file path to checkpoint given a iteration number"""
+        filename = "{}.{}".format(CHECKPOINT_FILENAME, iteration)
+        checkpoint_path = os.path.join(self.checkpoint_dir, filename)
+        return checkpoint_path
+
+    def start(self):
+        """Starts a background thread that deletes old checkpoints
```

Comment:
PEP 8 dictates that the docstring fgor methods read as a command, not a description:

```
"""Start background thread that ...
```

See more here: https://www.python.org/dev/peps/pep-0257/

Reply:
I didn't know this. Good to know.